pkg/render/gateway_api.go

```diff
@@ -0,0 +1,526 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package render
+
+import (
+	_ "embed"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/go-logr/logr"
+	v3 "github.com/tigera/api/pkg/apis/projectcalico/v3"
+	operatorv1 "github.com/tigera/operator/api/v1"
+	"github.com/tigera/operator/pkg/components"
+	rcomp "github.com/tigera/operator/pkg/render/common/components"
+	rmeta "github.com/tigera/operator/pkg/render/common/meta"
+	"github.com/tigera/operator/pkg/render/common/networkpolicy"
+	"github.com/tigera/operator/pkg/render/common/secret"
+	appsv1 "k8s.io/api/apps/v1"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	apiextenv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml" // gopkg.in/yaml.v2 didn't parse all the fields but this package did
+)
+
+var (
+	//go:embed gateway_api_resources.yaml
+	gatewayAPIResourcesYAML string
+)
+
+type yamlKind struct {
+	APIVersion string `yaml:"apiVersion"`
+	Kind       string `yaml:"kind"`
+}
+
+// This struct defines all of the resources that we expect to read from the rendered Envoy Gateway
+// helm chart (as of v1.1.2).
+type gatewayAPIResources struct {
+	namespace                 *corev1.Namespace
+	k8sCRDs                   []*apiextenv1.CustomResourceDefinition
+	envoyCRDs                 []*apiextenv1.CustomResourceDefinition
+	controllerServiceAccount  *corev1.ServiceAccount
+	envoyGatewayConfig        *corev1.ConfigMap
+	clusterRole               *rbacv1.ClusterRole
+	clusterRoleBinding        *rbacv1.ClusterRoleBinding
+	role                      *rbacv1.Role
+	roleBinding               *rbacv1.RoleBinding
+	leaderElectionRole        *rbacv1.Role
+	leaderElectionRoleBinding *rbacv1.RoleBinding
+	controllerService         *corev1.Service
+	controllerDeployment      *appsv1.Deployment
+	certgenServiceAccount     *corev1.ServiceAccount
+	certgenRole               *rbacv1.Role
+	certgenRoleBinding        *rbacv1.RoleBinding
+	certgenJob                *batchv1.Job
+}
+
+const (
+	GatewayAPIName                      = "gateway-api"
+	GatewayControllerLabel              = GatewayAPIName + "-controller"
+	EnvoyGatewayConfigName              = "envoy-gateway-config"
+	EnvoyGatewayConfigKey               = "envoy-gateway.yaml"
+	EnvoyGatewayDeploymentContainerName = "envoy-gateway"
+	EnvoyGatewayJobContainerName        = "envoy-gateway-certgen"
+)
+
+func GatewayAPIResourcesGetter() func() *gatewayAPIResources {
```

Comment:
This doesn't seem like something we need to have happen at run time since it will never change but it's redone multiple times when reconciliation happens. This seems to provide an unneeded source of panics that could be caught at build time.

Instead, could we just generate the structures here from the yaml at build time and save them to a file, similar to how we generate the version structures https://github.com/tigera/operator/tree/master/hack/gen-versions using go template?

Reply:
I think it's already doing what you think it should do.  In other words, it only reads and parses the YAML once. The only reason for the slightly odd (to you) form is to have the `resources` cache and associated lock internal to the function, instead of being globals that other code might accidentally misuse.